File: testdata/fixtures/go/safe/path_safe.go
package safe

import (
	"net/http"
	"path/filepath"
	"strings"
)

// SAFE: Path traversal prevention using filepath.Clean and strings.HasPrefix.
// Should NOT trigger GTSS-TRV-001.

const baseDir = "/var/www/static"

func HandleFileDownload(w http.ResponseWriter, r *http.Request) {
	filename := r.URL.Query().Get("file")

	// Clean the path to resolve ".." sequences
	cleaned := filepath.Clean(filename)

	// Build full path
	fullPath := filepath.Join(baseDir, cleaned)

	// Resolve to absolute path
	absPath, err := filepath.Abs(fullPath)
	if err != nil {
		http.Error(w, "invalid path", http.StatusBadRequest)
		return
	}

	// Verify the resolved path stays within the base directory
	if !strings.HasPrefix(absPath, baseDir) {
		http.Error(w, "access denied", http.StatusForbidden)
		return
	}

	http.ServeFile(w, r, absPath)
}

File: testdata/fixtures/javascript/safe/path_safe.ts
import { Request, Response } from 'express';
import * as fs from 'fs';
import * as path from 'path';

const ALLOWED_DIR = path.resolve(__dirname, 'public', 'files');

// SAFE: Path traversal prevention with resolve + startsWith check
export function servePublicFile(req: Request, res: Response): void {
  const fileName = req.params.file;

  // SAFE: Resolve the full path and verify it stays within the allowed directory
  const resolvedPath = path.resolve(ALLOWED_DIR, fileName);

  if (!resolvedPath.startsWith(ALLOWED_DIR)) {
    res.status(403).json({ error: 'Access denied' });
    return;
  }

  // Check file exists
  if (!fs.existsSync(resolvedPath)) {
    res.status(404).json({ error: 'File not found' });
    return;
  }

  res.sendFile(resolvedPath);
}

// SAFE: Using sendFile with root option and sanitized filename
export function downloadDocument(req: Request, res: Response): void {
  const fileName = req.params.filename;

  // SAFE: Strip path separators and null bytes from filename
  const sanitizedName = path.basename(fileName).replace(/\0/g, '');

  // Validate against allowlist of extensions
  const allowedExtensions = ['.pdf', '.txt', '.csv'];
  const ext = path.extname(sanitizedName).toLowerCase();

  if (!allowedExtensions.includes(ext)) {
    res.status(400).json({ error: 'File type not allowed' });
    return;
  }

  // SAFE: Use root option to restrict to specific directory
  res.sendFile(sanitizedName, { root: path.resolve(__dirname, 'documents') });
}
